Add select/deselect all buttons in download options panel in cart tab (Issue #267).

juzen2003 opened the pull request:
- Fixes #267, #770, #810, #388, #764, #728 
- Local branch: issues_in_product_type_selections

juzen2003 committed:
- This is a commit before switching branches.
- In cart/views.py, in line 661, add category_name as one of returned data,
  the data will be stored in the data-category later.
- In cart/cart.html, add category_name in data-category for #product_types
  and select-all, deselect-all buttons for each product type.
- In cart.js, add event handler for clicking select-all and deselect-all,
  they are still under experiment.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into issues_in_product_type_selections

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into issues_in_product_type_selections

juzen2003 committed:
Add select/deselect all buttons for file types in download options panel.
- In cart.html (line 9-12, 20-23), add select/deselect all buttons in
  download options panel for each file types.
- In cart.js (line 357-358), disable "deselect all" buttons when cart tab
  is first loaded because all checkboxes are checked at this time.
- In cart.js (line 77-120), update and add event handlers when users click
  checkboxes, click select/deselect all buttons for each product type, and
  click select/deselect all options for all product types.
- In cart.js (line 149-221), add functions: "updateDownloadFileInfo",
  "updateCheckboxes", "updateSelectDeselectBtn", and "isAllOptionStatusTheSame".
  These functions are used in above mentioned event handlers.
  - "updateDownloadFileInfo": update file info after selecting/deselecting
    download options.
  - "updateCheckboxes": Update checkboxes and disable/enable select all &
    deselect all buttons correspondingly.
  - "updateSelectDeselectBtn": Enable/disable select all and deselect all
    buttons.
  - "isAllOptionStatusTheSame": Check if all download options of a product
    type are the same, return true if all options are selected (when checked
    is true) or if all options are deselected (when checked is false).

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -74,17 +74,51 @@ var o_cart = {
-                }
-                o_cart.hideDownloadSizeSpinner(info.total_download_size_pretty);
-            });
+        $("#cart").on("click","#download_options input", function(e) {
```
Add a space after "click"

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -74,17 +74,51 @@ var o_cart = {
+            let productList = $(e.target).closest("ul");
+            let productInputs = productList.find("input");
+            let prodTypeSelectAllBtn = productList.prev("div").children().first();
+            let prodTypeDeselectAllBtn = productList.prev("div").children().last();
```
This whole thing seems very fragile to me. A slight change to the HTML will break it. Instead, since we have a category on the "ul" and a category on each of the select/deselect all buttons, why not retrieve the category from the "ul" and use it to search by "data" tag and class name for the two buttons?

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -653,11 +654,12 @@ def _get_download_info(product_types, session_id):
             pretty_name = 'Diagram Products'
         else:
             pretty_name = category + '-Specific Products'
+            category_name = 'specific'
```
I don't understand the need for a separate "category_name" here. First of all, it's possible to have more than one "X-Specific Products" in the list, like if you have both Voyager and Cassini observations in the cart at the same time. Your current code will use the same category_name for these two cases, which I assume could cause problems later. Instead, the "pretty_name" is already unique, and you could use it instead. If you need it to be a valid URL component/slug name/whatever (without spaces, "-", etc), then in the Django template you can just use "product_cat|slugify".

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -112,6 +146,79 @@ var o_cart = {
+            for (let checkbox of checkboxesOptions) {
+                $(checkbox).prop("checked", false);
+            }
+        }
```
else.... log an error?

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -112,6 +146,79 @@ var o_cart = {
+    updateSelectDeselectBtn: function(isAllSelected, isAllDeselected, selectBtn, deselectBtn) {
+        /**
+         * Enable/disable select all and deselect all buttons based on
+         * parameters allSelected & allDeselected passed in.
```
Actually the parameters are named isAllSelected etc. Also more clear to say "Enable/disable one pair of select all and deselect all buttons..."

rfrenchseti commented on opus/application/apps/cart/templates/cart/cart.html:
```diff
@@ -6,14 +6,22 @@ <h1>Zipped File Links</h1>
         <!-- <p>Total Size (before zip): <span id="op-total-download-size">{{ total_download_size_pretty }}</span></p> -->
         <p class="op-total-size">Total Size (before zip): <span class="spinner">&nbsp;</span><span id="op-total-download-size">{{ total_download_size_pretty }}</span></p>
 
         <div id="download_options">
```
Now would probably be a fine time to rename this to "op-download-options" since we're using it everywhere.

rfrenchseti commented on opus/application/apps/cart/templates/cart/cart.html:
```diff
@@ -6,14 +6,22 @@ <h1>Zipped File Links</h1>
         <!-- <p>Total Files: <span id="total_files">{{ download_count }}</span></p> -->
         <!-- <p>Total Size (before zip): <span id="op-total-download-size">{{ total_download_size_pretty }}</span></p> -->
         <p class="op-total-size">Total Size (before zip): <span class="spinner">&nbsp;</span><span id="op-total-download-size">{{ total_download_size_pretty }}</span></p>
 
```
I think it would help to add a phrase like "Select which product types to include in downloads:" after the size and before the "Select all product types" buttons.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -74,17 +74,51 @@ var o_cart = {
+                                           ".op-cart-select-all-btn", ".op-cart-deselect-all-btn");
+        });
+
+        // Event handler when clicking "Select all options" and "Deselect all options" buttons.
```
Make sure to change comment based on new button names

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -112,6 +146,79 @@ var o_cart = {
+         */
+        if (($(target).hasClass("op-cart-select-btn") ||
+             $(target).hasClass("op-cart-select-all-btn"))) {
+            for (let checkbox of checkboxesOptions) {
```
Since "checkboxesOptions" is the result of a jQuery find(), can't you just do `checkboxesOptions.prop("checked", true);` and it will automatically loop through them all? If not, change the "let" to "const".

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -112,6 +146,79 @@ var o_cart = {
+        if (productList.length === 1) {
+            return checked === productList.is(":checked");
+        }
+        for (let productOption of productList) {
```
Use const

rfrenchseti reviewed (changes requested):
- The "Select all options" and "Deselect all options" buttons need to be below the "Total Size" line. They should also be called "Select all product types".
- I think it would look better if the Select all/Deselect all buttons were a little to the right, perhaps lined up with the (i) Text instead of with the checkbox.

juzen2003 committed:
Update based on github review comments.
- In view.py (line 645,655,660), remove category_name (not used).
- In cart.html (line 12-24), set product_cat|slugify in data-category for
  each ul and pair of select/deselect all buttons. Move "Select all product types"
  and "deselect all product types" below total size.
- In cart.js (line 172, 175), remove for loop and use
  $(checkboxesOptions).prop("checked", true/false); directly.
- In cart.js (line 182,185,187,190), rename isAllSelected to enableSelectAll,
  and isAllDeselected to enableDeselectAll.
- In cart.js (line 210), change for (let productOption of productList) to
  for (const productOption of productList).

juzen2003 committed:
Add total files count below total size in download options panel.
- In cart.html (line 12), add a p tag to display total download counts.
  It's similar to the p tag for total download size.
- In opus.css (1416-1417), add css styling for .op-total-download .spinner
  (same as the one for .op-total-size .spinner).
- In cart.js (line 43,554,565), rename "downloadSizeSpinnerTimer" to
  "downloadSpinnerTimer", "showDownloadSizeSpinner" to "showDownloadSpinner",
  and "hideDownloadSizeSpinner" to "hideDownloadSpinner".
- In cart.js (line 565), update hideDownloadSpinner to take in downloadCount
  as the 2nd parameter. Calling this function will display both total size
  and total files after removing spinners when api return comes back.

juzen2003 committed:
Put a table (init draft) to display info on download options panel. Event
handler functions in download options panel are not updated yet (this is
a commit before switching branch).
- In opus.css (line 891-899), add styling for table in download options.
  Remove vertical border, and increase the width for table in download
  options panel.
- In cart.html (line 8-85), update structure to display info in download
  options panel as a table (initial draft).
- Copy and backup original cart.html to cart_ul_backup.html. (this will
  be removed later in the future, just a backup for now).

juzen2003 committed:
Tune the border of table in download options panel.
- In opus.css (line 1624), make the css selector more specific so that
  the styling here will not affect table border in download options panel.
- In cart.html (lin 33-47), this is the draft of the table (still in
  progress).

juzen2003 committed:
Update download options panel with a table of product types. Make download
options header stays at the top and only product types area is scrollable.
- In base.html (line 260), set the background of download options card body
  to white so that the gray product types subheader are visible.
- In cart.html (line 8, 24-25), rename "op-download-header" to
  "op-download-options-header", "op-download-options-panel" to
  "op-download-options-product-types", and "op-download-table" to
  "op-download-options-table".
- In cart.html (line 16-21, 37-41), put "v" as select all and "x" as
  deselect all buttons for each product type, and also add them next to
  select all product types and deselect all product types buttons.
- In opus.css (line 896-924), add styling for new download options panel
  changes.
- In cart.js (line 35-351), init perfect scrollbar when
  .op-download-options-product-types is rendered.
- In cart.js (line 280, 288, 290-292), update adjustProductInfoHeight to
  get the correct height for download options table, this is required to
  make only product types area scrollable.

juzen2003 committed:
Fix event handler functions for checkboxes and select/deselect buttons in
download options. This will make sure the behavior of clicking checkboxes and
select/deselect buttons are correct.
- In cart.html (line 57), add data-category to input checkboxes in download
  options.
- In cart.js (line 110-146, 238), update event handler functions for checkboxes
  and select/deselect buttons, and also update getDownloadFiltersChecked.
  This will ensure the correct behavior of clicking checkboxes and buttons
  in download options.

juzen2003 committed:
Fix the scrollbar height issue in download options and move "v" and "x"
buttons ahead of each product category.
- In cart.html (line 38,41), move "v" and "x" buttons ahead of each product
  category in download options.
- In cart.js (line 283, 289-291), update adjustProductInfoHeight to have
  the correct scrolling height for download options panel & sidebar.

juzen2003 committed:
Update the layout of download options panel, move the "Download Options"
to panel title when panel is available.
- In cart.js (line 289-290, 295), update adjustProductInfoHeight to get
  the correct height of scrolling area for download options table in both
  pane and panel.
- In cart.js (line 226-233), update displayCartLeftPane to move "Download
  Options" to panel header when panel is available.
- In opus.css (line 1196-1201), styling for download options panel header.

juzen2003 committed:
Increase the width of download options area so that the table looks better.
- In opus.css (line 949-951, 1060), increase the width of download options
  area from 28em to 36em. Also modify the width of #cart .op-cart-nav-container
  to make slider takes more space at the 1st line when screen is narrower.

rfrenchseti reviewed (changes requested):
Good start!

- The "Total files" shouldn't say "(before zip)" because the number of files doesn't change with zipping. It would also be nice to have the total size and total files on the same line to save vertical space.
- I think the buttons for "select all product types" and "deselect all product types" are pretty ugly. There's no reason to have them as wide as the whole options pane. Please come up with some ideas to make them look smaller and more aesthetically pleasing.
- The columns for # Obs, # Files, and Size should have their contents right justified
- The values for # Obs and # Files should have thousand-commas added
- I don't think there should be so much space between the checkmark, X, and product category. I also think the icons might look better smaller, but I'd have to see them to be sure. It would be more intuitive if the checkmark and X were basically over the checkbox column, rather than off to the side. Given the checkbox and the space to the right of the checkboxes, there should be enough space in the product category header to have both a checkmark and X over that region.
- I don't think the product names need to be in bold
- Now that we have only the table scrollable, it's possible to put in enough zip files that you can't see the products at all, and there's no way to scroll down to see them. I think with this design we have to implement the "Download Links" pop-out window at the same time (#764).
- The table scrolling includes the table header. The header should stay put and the scrolling should only affect the product types and counts.
- Make the heading "# Obs" and "# Files" so it fits on one header line.
- When the browser is narrow and the download options pane is a slide-out, its width is responsive to the browser width. This wasn't a big deal before, but with the new table layout it can look really bad. The pane slide-out should be a constant width, or at least have a maximum width that's the same as the non-slide-out pane.
- I haven't reviewed any of the code changes.

juzen2003 committed:
Remove (before zip) after "Total Files" in download options pane.
- In cart.html (line 12-19), change "Total Files (before zip)" to "Total
  Files".

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into issues_in_product_type_selections

juzen2003 committed:
Put "Total Size" and "Total Files" into one line in Download Options.
- In cart.html (line 12-21), put both .op-total-size and .op-total-download
  into a div. We will set both p tags to display inline so that they will
  stay in one line.
- In opus.css (line 935-938), set .op-total-size and .op-total-download
  to have display: inline; so that they will stay in one line.

juzen2003 committed:
Update the layout of download options pane. Use boostrap buttons for
select all product types and deselect all product types buttons, and add
some styling tuning to download options table.
- In opus.css (line 921-924), add styling to buttons in download options
  to make when buttons are disabled, there is no cursor pointer.
- In cart.html (line 22, 24, 27), add bootstrap button classes to select
  all product types and deselect all product types buttons. Also add some
  margins in the area.
- In cart.html (line 38-40, 80-82), add boostrap text class to make header
  and contents of # Obs, # Files, and Size to align to the right.
- In cart.html (line 73), remove bootstrap font weight class so that names
  of product types are not in bold.

juzen2003 committed:
Add commas to total files, # files, and #obs in download options using
django template filter from django.contrib.humanize.
- In settings.py (line 119), add django.contrib.humanize to INSTALLED_APPS.
- In cart.html (line 20, 81-82), use intcomma filters to have commas displayed
  in those outputs.

juzen2003 committed:
Convert download options table to a div table using bootstrap row and col
classes so that only the table body is the scrollable area for perfect
scrollbar.
- In cart.html (line 33-72), modify the table tag to div table using bootstrap
  row and col classes.
- In opus.css (line 924-932), styling for perfect scrollbar on div table
  (.op-product-type-table-body).
- In cart.js (line 362), change the ps container to .op-product-type-table-body.
  This is the table body container for div table.
- In cart.js (line 290-302), modify adjustProductInfoHeight to make sure
  ps container height is calculated correctly for new div table body.
- In cart.js (line 161-162, 319-320), modify to pass in total_download_count
  with commas to hideDownloadSpinner. This will make sure number displayed
  in total files has comma.

juzen2003 committed:
Fix the width of download options panel. Now the width of the panel and
its contents will not extend when we are increase the width of the browser.
- In base.html (line 255), remove the w-75 class so that width of download
  options panel will not extend while we are increasing the browser width.
- In base.html (line 260), put pr-0 so that the right space next to product
  type table in the panel is removed. This will look more like the regular
  download options pane.
- In opus.css (line 1212-1214), fix the width of #op-cart-download-panel
  (download options panel).

juzen2003 committed:
Fix the scrollable height of table contents in download options panel.
- In cart.js (line 297-298), fix the calculation of scrollable contents
  height. This will make sure we have correct scrollable height for table
  contents when in download options panel.

juzen2003 committed:
Add initial draft of lower right popover window for download links in
cart page.
- In cart.html (line 3) and base.html (line 463-471), remove original
  div for download links in cart.html and move them to base.html (div
  #op-download-links).
- In footer.html (line 7-9), add a button .op-download-links-btn to open
  up popover window for download links in cart page at the lower right
  corner of the footer area.
- In opus.css (line 1651-1654), styling for .op-download-links-btn to make
  sure it shows up at lower right corner.
- In cart.js (line 264-294), modify downloadZip function to dynamically
  update and display contents in popover window with latest download links.
- In cart.js (line line 148-154), initialize popover.

juzen2003 committed:
Modify the style of popover window and button. Remove error messages from
popover window, and put them into a modal.
- In footer.html (line 7-8), update the styling for lower right popover
  button.
- In base.html (line 435-444), add #op-download-links-error-msg modal for
  errors of download links. Now a modal will pop up when there is any error
  returned after clicking download buttons.
- In cart.js (line 148-167), make sure popover are only triggered by manual
  event, and we will set the event handler manually on buttons to open/close
  popover. The reason we do this is to make sure when popover is manully
  open by show method, it will close by clicking the button once only. If
  we didn't set the trigger option manual and add event handler manually,
  we need to click the button twice to close a manually open popover.
- In cart.js (line 276-303). modify downloadZip to open a modal when download
  api returns an error. Also, the popover button is added only when the
  download api has no error or a download history exists.

juzen2003 committed:
Update the style of popover window for download links history. Add a "x"
button to close the popover window. Fix the popover jumping issue when
spinner is running.
- In cart.js (line 154-156), add a "x" button in the popover window. This
  is used as a close button for popover window.
- In cart.js (line 173-177), add the event handler to manually close popover
  window when user clicks "x" button.
- In cart.js (line 337-351), add enablePSinDownloadLinksWindow function.
  This is used to enable and update PS when the popover window has reach
  to its max height.
- In cart.js (line 314-322), update downloadZip to provide a max height
  to popover window and enable ps when necessary.
- In cart.js (line 290-291), update popover's position when spinner is running.
  This will fix the popover jumping issue when displaying the spinner.
- In opus.css (line 970-983), add the styling for download links popover
  window, including the styling for "x" button and ps in the window.
- In footer.html(line 7-8), remove title attribute in button tag, we will
  set the popover title when initializing it.
- In base.html (line 476), rename "zippedFiles" to "op-zipped-files".

juzen2003 committed:
Update the threshold point for enabling the download options slide-out
panel. (Change from 940px to 1160px).
- In cart.js (line 11), change cartLeftPaneThreshold from 940 to 1160.
- In opus.css (line 1179, 1212), change the threshold from 940px to 1160px,
  and 941px to 1161 px.

juzen2003 committed:
Update code style and comments in cart.js.
- In cart.js (line 121-134), move the code snippets from line 74 to here.
- In cart.js (line 135), update comments.
- In cart.js (line 208-209), log an error if the button in download options
  pane has the wrong class.

juzen2003 committed:
Fix the jumping issue of download links popover window when spinner is running.
- In cart.js (line 306, 331), update the popover window position after
  hiding the spinner.

juzen2003 committed:
Change download history button to a blue link, remove the little oval white
background of "x" button in popover window.
- In footer.html (line 7-9), change button tag to a tag for download history
  button.
- In opus.css (line 970-983), add styling for download history blue link
  in the footer area.
- In opus.css (line 988-996), remove the white oval background of "x" in
  the popover window and make it the same color as the popover header background.

juzen2003 committed:
Add a popover footer and a clear all button to clear download links history.
- In opus.css (line 999-1010), add styling for popover footer. It's copied
  and modified based on bootstrap .modal-footer to create a pretty footer
  for popover window.
- In cart.js (line 377-388), create a function clearDownloadLinksHistory.
  This is to remove all download links history and display a message about
  the empty history to the user.
- In cart.js (line 196-198), attached clearDownloadLinksHistory as the
  event handler function for clicking clear all button in popover window.
- In cart.js (line 341, 345), remove the empty history message and disable
  clear all button when there is a valid download link.

juzen2003 committed:
Update downloadLinksMaxHeight to 250 and set the init state of .op-clear-history-btn
to disabled.
- In cart.js (line 13, 28), update downloadLinksMaxHeight to 250 and set
  init state of .op-clear-history-btn to disabled.

juzen2003 committed:
Update popover window for download links history. Make "Download History"
button always visible but disabled when there is no history. Also make the
browser size modal window display in front of all other windows (higher
z-index).
- In footer.html (line 7), add is-disabled class to Download History button
  to grey out the button.
- In footer.html (line 6), rename to "Planetary Ring-Moon Systems Node" to
  "PDS Ring-Moon Systems Node".
- In opus.css (line 976-981), create .is-disabled with the styling for
  disabled a tag.
- In opus.css (line 1027-1036), add z-index styling to make sure browser
  size window is displayed on top of all other windows, and popover window
  is covered by the modal backdrop.
- In cart.js (line 30-35, 176), create downloadLinksPopoverTitle to store
  title for popover window.
- In cart.js (line 350, 392), proper enable/disable download history button.
- In cart.js (line 13), change downloadLinksMaxHeight to 200, this will
  make the popover window shorter.

juzen2003 committed:
Rename "is-disabled" to "op-a-tag-btn-disabled", "downloadLinksMaxHeight"
to "downloadLinksPBMaxHeight".
- In footer.html (line 7), opus.css (line 977), cart.js (12-13, 344, 350,
  372, 392), rename "is-disabled" to "op-a-tag-btn-disabled", "downloadLinksMaxHeight"
  to "downloadLinksPBMaxHeight".

juzen2003 committed:
Update the if condition logic in isAllOptionStatusTheSame function in cart.js
- In cart.js (line 270), change if condition from "=== !checked" to "!== checked"
  to improve the readability.

rfrenchseti commented on opus/application/apps/cart/templates/cart/cart.html:
```diff
@@ -1,41 +1,70 @@
-            <span class="spinner">&nbsp;</span>
+        <div class="op-download-options-header border-bottom border-light">
+            <h1>Download Options</h1>
+            <!-- <p>Total Files: <span id="total_files">{{ download_count }}</span></p> -->
```
Remove comments

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -368,10 +368,10 @@ ul.op-search-menu {
 }
 
-#cart .fa-times {
+/* #cart .fa-times {
```
Remove comments

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -889,8 +889,45 @@ a:-webkit-any-link, a::any-link {
+panel */
+#cart .op-cart-select-btn:hover, #cart .op-cart-deselect-btn:hover,
+#cart .op-cart-select-btn:focus, #cart .op-cart-deselect-btn:focus {
+    background-color: #F0F3F5 !important;
```
Is this really the best way to do this? Isn't there some way to simply disable the hover action entirely? Setting the background to a hardcoded color seems like a hack.

juzen2003 replied:
In our case, we don't want to disable the entire hover & focus effect. We only want to remove the outline and background color when we mouse over to "v" and "x". The background color of the hover effect is coming from button.bg-light:hover (from btn-light:hover if we change the class in html), and the easiest way to bypass this is to customize css for these two buttons. I've modified the styling to "transparent" (not a specific hardcoded color now) so the hover effect background color will just match to its parent container which is still a .bg-light from table sub-header.

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -8,7 +8,31 @@
+                                     "<div class='popover-body'></div>" +
+                                     "<div class='popover-footer'>" +
+                                     "<button class='op-clear-history-btn btn btn-sm btn-secondary'" +
+                                     "type='button' title='Clear all history' disabled>Clear all</button></div>" +
```
In other dialogs (like Select Metadata), every word in the button is capitalized, so to be consistent this should be "Clear All"

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -110,6 +117,161 @@ var o_cart = {
+                    $(target).hasClass("op-cart-deselect-all-btn"))) {
+            $(checkboxesOptions).prop("checked", false);
+        } else {
+            opus.logError(`Target button in download options left pane has the wrong class.`);
```
Should use "..."

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -110,6 +117,161 @@ var o_cart = {
+         * return true if all options are selected (when checked is true)
+         * or if all options are deselected (when checked is false). This
+         * needs to be executed after updateCheckboxes (wait until all
+         * checkboxes properties are updated)
```
Add period at end

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -155,49 +326,98 @@ var o_cart = {
                 } else {
-                    $(`<li><a href = "${data.filename}" download>${data.filename}</a></li>`).hide().prependTo("ul.zippedFiles", "#cart_summary").slideDown("slow");
+                    // To dynamically update and display contents of an open popover, we have to make sure html
+                    // in both (1) #op-download-links ( content when popover is initialized) and (2) .popover-body
```
Remove space after (

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -155,49 +326,98 @@ var o_cart = {
+                    // in both (1) #op-download-links ( content when popover is initialized) and (2) .popover-body
+                    // (when popover is open) are synced up with updates. To achieve this, we have to call show
+                    // method from popover to update content, and make sure the selector managing DOM are selecting
+                    // the same elements in both #op-download-links and .popover-body. (lenght === 2).
```
length

rfrenchseti reviewed (changes requested):
- The horizontal line under the table header goes all the way to the right edge of the pane, while the line above the header stops at the end of the words. I think the top line has the proper behavior.
- Also how about changing the tooltip to "Select/Deslect all in category" to make it clear how it's different from the bit select/deselect all buttons (edited)

juzen2003 committed:
Update the style of download options pane and clean up some comments and
code style in cart.js, cart.html, opus.css.
- In cart.html (line 27), change class pr-3 to mr-3 so that the border below
  download options table header will stop at the end of the words in the
  header.
- In cart.html (line 38-40), update the tooltip for "v" and "x" to "Select
  all in category" and "Deselect all in category".
- In cart.html (line 6), opus.css (line 371), remove unwanted comments.
- In cart.js (line 28), change "Clear all" to "Clear All".
- In cart.js (line 234), use "" for a string.
- In cart.js (line 264, 336, 339), clean up comments.
- In opus.css (line 901-902), update the styling to remove the hover effect
  for .op-cart-select-btn and .op-cart-deselect-btn.
